### PR TITLE
Migrate more properties to lazy array

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | refacto
| Category?         | FO
| BC breaks?        | yes - but nothing serious, if somebody is accessing these properties in a template, he won't notice a difference.
| Deprecations?     | yes
| How to test?      | Tests should be sufficient to show any error.
| UI Tests          | 🟢 https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/7046752953
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | 

### Description
- Migrates another batch of easy properties to lazy array.
- Enables passing presented Product object into getProductProperties without manually assigning id_product.
- Warms up tax rules group and pack properties so they don't have to be manually fetched.

### Performance with 24 products per category
Average load time - 470 ms ➡️ 438 ms
Query count - 636 ➡️ 538

### CODE REVIEW

**Review by matthieu-rolland — COMMENTED** [review]
there are many tests failing in the UI suite you ran... maybe you should restart the failed tests just in case something has been fixed since then and clear anything that's not linked to your PR

**Comment by Hlavtox** [comment]
@matthieu-rolland I will check the tests one by one and fix it. ;-)

**Comment by Hlavtox** [comment]
@matthieu-rolland Restarting the tests fixed all of them with one exception, but it's about api secrets. 👍

**Review by matthieu-rolland — DISMISSED** [review]

File: classes/Product.php [hunk 1]
```diff
@@ -1692,6 +1692,16 @@ public static function getSimpleProducts($id_lang, Context $context = null)
      * @return bool
      */
     public function isNew()
+    {
+        return $this->isNewStatic((int) $this->id);
+    }
```
Comment (PululuK) [inline classes/Product.php]:
```suggestion
    {
        return static::isNewStatic((int) $this->id);
    }
```

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 2]
```diff
@@ -462,6 +500,90 @@ public function getEcotax()
...
+    {
+        if (!isset($this->product['category_name'])) {
+            $categoryName = (string) Db::getInstance()->getValue(
+                'SELECT name FROM ' . _DB_PREFIX_ . 'category_lang 
+                WHERE id_shop = ' . (int) Context::getContext()->shop->id
+                . ' AND id_lang = ' . (int) $this->language->id
+                . ' AND id_category = ' . (int) $this->product['id_category_default']
+            );
+            $this->product['category_name'] = !empty($categoryName) ? $categoryName : null;
+        }
```
Comment (PululuK) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:559]:
```suggestion
        if (!isset($this->product['category_name'])) {
            $categoryObject = new Category((int) $this->product['id_category_default'], null, (int) Context::getContext()->shop->id);
            $this->product['category_name'] = $categoryObject->getName((int) $this->language->id);
        }
```

### NOTE :
 I suggest to add a static cache for category object

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 3]
```diff
@@ -328,9 +348,27 @@ public function getFileSizeFormatted()
...
+
+            /*
+             * There is an optional cache_has_attachments property, which if passed, informs us if the product
+             * has attachments or not in a fast way. We will load attachments only if this property was not passed
+             * or is true.
+             *
+             * This property which needs to be managed every time a file is changed.
+             * It can sometimes lead to database inconsistency.
+             */
+            if (!isset($this->product['cache_has_attachments']) || $this->product['cache_has_attachments']) {
```
Comment (PululuK) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:364]:
```suggestion
            if (!empty($this->product['cache_has_attachments'])) {
```

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 4]
```diff
@@ -462,6 +500,90 @@ public function getEcotax()
...
+            if (!empty($this->product['id_manufacturer'])) {
+                $manufacturerName = Manufacturer::getNameById((int) $this->product['id_manufacturer']);
+                if (!empty($manufacturerName)) {
+                    $this->product['manufacturer_name'] = $manufacturerName;
+                }
+            }
+        }
+
+        return $this->product['manufacturer_name'];
+    }
```
Comment (PululuK) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:525]: ### Suggestion : 
rather improve the `Manufacturer::getNameById` to return a null when the param `$idManufacturer` provided is not valid.

```suggestion
    public function getManufacturerName()
    {
        if (!isset($this->product['manufacturer_name'])) {
            $this->product['manufacturer_name'] = Manufacturer::getNameById((int) $this->product['id_manufacturer']);
        }

        return $this->product['manufacturer_name'];
    }
```

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 5]
```diff
@@ -462,6 +500,90 @@ public function getEcotax()
...
+    {
+        if (!isset($this->product['category'])) {
+            $categoryLinkRewrite = Category::getLinkRewrite(
+                (int) $this->product['id_category_default'], (int) $this->language->id
+            );
+            $this->product['category'] = !empty($categoryLinkRewrite) ? $categoryLinkRewrite : null;
+        }
+
+        return $this->product['category'];
+    }
```
Comment (PululuK) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:542]: hummm maybe rename the function to `getCategoryLinkRewrite` ?

**Review by PululuK — COMMENTED** [review]
Thanks @Hlavtox  ! Small feedbacks and LGTM :)

File: src/Adapter/Presenter/Product/ProductLazyArray.php (hunk 5 above)
Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:542]: That would be better, but it would be a BC break, because you could no longer access it by $product['category'], is it worth it?

File: src/Adapter/Presenter/Product/ProductLazyArray.php (hunk 3 above)
Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:364]: This would change the behavior, that's why I kept the current logic.

`if (!isset($this->product['cache_has_attachments']) || $this->product['cache_has_attachments']) {`
This means `if we don't know (it wasn't passed)` OR `we know` that product has attributes, load them.

`if (!empty($this->product['cache_has_attachments'])) {`
This means `if we know`, that product has attributes, load them.

I want to make it as easy to use as possible.

File: classes/Product.php (hunk 1 above)
Comment (Hlavtox) [reply classes/Product.php]: Applied :-)

**Comment by matks** [comment]
@Hlavtox git conflicts 😭

File: src/Adapter/Presenter/Product/ProductLazyArray.php (hunk 2 above)
Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:559]: Well, we could do this, but there is a slight additional performance overhead, checking if category image exists, whole object construct etc.

I checked it in the current form and the Category objects are not constructed during product listings. So if somebody uses this property, it would be the first time.

I thought that maybe constructing the Category object would be good if we could reuse it... but the only place we need categories in product listings is for URL rewrite. And the rewrite is using `Category::getLinkRewrite` method, which has it's own caching and is faster than constructing Category objects.

TLDR - I think it's good, but only if we NEEDED the category for 2 or more properties.

**Comment by Hlavtox** [comment]
@PululuK @matks @matthieu-rolland Conflicts resolved

**Review by M0rgan01 — APPROVED** [review]

**Review by ShaiMagal — APPROVED** [review]

**Comment by Hlavtox** [comment]
Ping @PrestaShop/qa-automation for tests check, there is one failure, but it's the same as in other PRs. :-)

**Comment by nesrineabdmouleh** [comment]
Hello @Hlavtox 
The error is not related to the PR, I will fix it ASAP!

**Comment by Hlavtox** [comment]
@nesrineabdmouleh Actually I think it's the core were is the issue. 🤔

**Review by florine2623 — APPROVED** [review]
Hello @Hlavtox ,

I did some standard functional tests, LGTM.

It is QA ✅
